js/ui/windowManager.js

```diff
@@ -1832,3 +1887,522 @@ const WindowManager = new Lang.Class({
+
+    _removeSession : function(session) {
+        let idx = this._sessions.indexOf(session);
+        if (idx != -1) {
+            this._sessions.splice(idx, 1);
+            return true;
+        }
+        return false;
+    },
+
+    rotateInCompleted: function(actor) {
+        if (this._removeEffect(this._rotateInActors, actor)) {
+            Tweener.removeTweens(actor);
+            actor.opacity = 255;
+            actor.rotation_angle_y = 0;
+            actor.set_cull_back_face(false);
+        }
+    },
+
+    rotateOutCompleted: function(actor) {
+        if (this._removeEffect(this._rotateOutActors, actor)) {
+            Tweener.removeTweens(actor);
+            actor.hide();
+            actor.rotation_angle_y = 0;
+            actor.set_cull_back_face(false);
+        }
+    },
+
+    _getBuildManifestsAt: function(location) {
+        function generateArrayFromFunction(func) {
+	        let arr = [];
+	        let result;
+
+	        while ((result = func.apply(this, arguments)) !== null) {
+		        arr.push(result);
+	        }
+
+	        return arr;
+	    }
+
+        function listDirectory(directory) {
+	        let file = Gio.File.new_for_path(directory);
+	        let enumerator = file.enumerate_children('standard::name', 0, null);
+	        let directoryInfoList = generateArrayFromFunction(() => enumerator.next_file(null));
+	        return directoryInfoList.map(function(info) {
+	            return {
+	                name: info.get_name(),
+	                type: info.get_file_type()
+	            };
+	        });
+	    }
+
+        location = location + '/app/';
+        let manifests = [];
+        let apps = listDirectory(location);
+        return manifests = apps.map(function(app) {
+            return location + app.name + '/current/active/files/manifest.json';
+        });
+    },
+
+    /**
+     * _getManifestPath:
+     * @flatpakID: flatpak id of the app to get the manifest path for
+     *
+     * Looks for the manifest of the app in the user and
+     * system installation path for flatpaks.
+     */
+    _getManifestPath: function(flatpakID) {
+        function readFileContents(path) {
+            let cmdlineFile = Gio.File.new_for_path(path);
+            let [ok, contents, etag] = cmdlineFile.load_contents(null);
+            return contents;
+        }
+
+        let manifests = this._getBuildManifestsAt(Flatpak.Installation.new_user(null).get_path().get_path())
+            .concat(this._getBuildManifestsAt(Flatpak.Installation.new_system(null).get_path().get_path()));
+
+        for (let j = 0; j < manifests.length; j++) {
+            let manifest;
+            try {
+                manifest = JSON.parse(readFileContents(manifests[j]));
+            } catch(err) {
+                logError(err, ' No build manifest found at ' + manifests[j]);
+                continue;
+            }
+		    if (manifest.id == flatpakID)
+			    return manifests[j];
+        }
+        return null;
+    }
+});
```

Comment:
We had a quick chat about it in the standup. Cosimo is ok with it in the shell, the service might be a bit overengineering. I think it is ok for now.